rinja/src/values.rs

```diff
@@ -0,0 +1,66 @@
+#![allow(clippy::new_without_default)]
+
+#[cfg(feature = "alloc")]
+use alloc::boxed::Box;
+#[cfg(all(not(feature = "std"), feature = "alloc"))]
+use alloc::collections::BTreeMap as HashMap;
+#[cfg(feature = "alloc")]
+use alloc::string::String;
+use core::any::Any;
+#[cfg(feature = "std")]
+use std::collections::HashMap;
+
+/// Used to store values of any type at runtime.
+///
+/// If the `std` feature is enabled, it internally uses the [`HashMap`] type, if the `alloc`
+/// feature is enabled, it uses the [`BTreeMap`](alloc::collections::BTreeMap) type, otherwise
+/// it uses an empty type and does nothing.
+#[derive(Default)]
+pub struct Values(#[cfg(feature = "alloc")] HashMap<String, Box<dyn Any>>);
```

Comment:
I kinda dislike the use of a HashMap in here. Unless there are quiet a lot key-value pairs, a simple slice `&[(&str, &dyn Any)]` will be excessively much faster. My guess is that a HashMap with many intermediate allocations won't be faster for less than 20 values. (I did not benchmark my guess, though. :) )

Maybe `Values` could be a trait instead?

Reply:
Not using a map sounds good to me. I suppose by making `Values` a trait, we would let it up to users to implement it on whatever they want? But then the question is: should we implement on some types by default as well (like `&(&str, &dyn Any)` or `HashMap<String, Box<dyn Any>>`)?

Reply:
I think it would be useful to implement it for some types, so it works "out of the box". But I would not seal the trait.

Reply:
What do you mean by "seal the trait"?

Reply:
```rust
mod private { // ← not `pub`, no you can't implement this trait outside of the crate
    pub trait Sealed {}
}

pub mod MyTrait: private::Sealed {}
```

<https://predr.ag/blog/definitive-guide-to-sealed-traits-in-rust/#what-are-sealed-traits>

Reply:
Ah so making the trait impossible to implement outside of the crate. I didn't know there was a name for that. TIL, thanks!

Reply:
Big question: making the `Values` trait fully generic over the key creates a lot of issues:
 * DynTemplate is not dyn-compatible anymore
 * We get unsolvable errors like:
     ```
     4  | #[derive(Template)]
   |          ^^^^^^^^
   |          |
   |          expected `&K`, found `&str`
   |          arguments to this method are incorrect
    ```

Would it be ok to force the key to be `AsRef<str>`?

Reply:
Yes, I think it's a good idea to force the key to be str'ish. Maybe `K: Borrow<str>` would be the best solution? So the interface is the same as `HashMap::get()`, and because `Borrow<str>` implies that `K`'s hash is the same as the borrowed `str`'s hash.

Reply:
Sounds good to me.